src/messages.rs

```diff
@@ -194,49 +202,97 @@ impl HopMessage {
-        let bytes_to_sign = try!(serialise(&(&content, full_id.public_id())));
+        let sig = sign::sign_detached(&try!(serialise(&content)), full_id.signing_private_key());
         Ok(SignedMessage {
             content: content,
-            public_id: *full_id.public_id(),
-            signature: sign::sign_detached(&bytes_to_sign, full_id.signing_private_key()),
+            grp_lists: Vec::new(),
+            signatures: iter::once((*full_id.public_id(), sig)).collect(),
         })
     }
 
-    /// Confirms the signature against the claimed public ID.
+    /// Confirms the signatures.
     pub fn check_integrity(&self) -> Result<(), RoutingError> {
-        let signed_bytes = try!(serialise(&(&self.content, &self.public_id)));
-        if sign::verify_detached(&self.signature,
-                                 &signed_bytes,
-                                 self.public_id().signing_public_key()) {
-            Ok(())
-        } else {
-            Err(RoutingError::FailedSignature)
+        let signed_bytes = try!(serialise(&self.content));
+        for (pub_id, sig) in &self.signatures {
+            if !sign::verify_detached(sig, &signed_bytes, pub_id.signing_public_key()) {
+                return Err(RoutingError::FailedSignature);
+            }
+        }
+        if let Authority::Client { ref client_key, .. } = self.content.src {
+            if let Some(pub_id) = self.signatures.keys().next() {
+                if self.signatures.len() == 1 && pub_id.signing_public_key() == client_key {
+                    return Ok(());
+                }
+            }
+            return Err(RoutingError::FailedSignature);
+        }
+        Ok(())
+    }
+
+    /// Returns whether the message is signed by the given public ID.
+    pub fn signed_by(&self, pub_id: &PublicId) -> bool {
+        self.signatures.contains_key(pub_id)
+    }
+
+    /// Appends the group list to the message.
+    pub fn add_group_list(&mut self, group_list: GroupList) {
+        if self.content.src.is_client() {
+            return; // Clients are validated based on their names.
+        }
+        // TODO: We currently just add empty group lists in clients and joining nodes. Clients will
+        // need to know their proxy's group to verify messages, and joining nodes will have to
+        // store their new group lists from the beginning.
+        if self.grp_lists.is_empty() && !group_list.pub_ids.is_empty() {
+            // Drop signatures not validated by the first group list.
+            let invalid_ids = self.signatures
+                .keys()
+                .filter(|pub_id| !group_list.pub_ids.contains(pub_id))
+                .cloned()
+                .collect_vec();
+            for invalid_id in invalid_ids {
+                let _ = self.signatures.remove(&invalid_id);
+            }
+        }
+        self.grp_lists.push(group_list);
+    }
+
+    /// Adds the given signature, if it is valid and new.
```

Comment:
You mean because otherwise we _do_ serialise it every time we add a signature?
That's a good point. And maybe that's all premature optimisation. I'd say go with the more elegant solution, if in doubt. We can optimise later.